tests/JoinSqlTest.php

```diff
@@ -762,4 +763,230 @@ public function testJoinActualFieldNamesAndPrefix(): void
             ],
         ], $this->getDb());
     }
+
+    public function testJoinSavingForeignCustomIdFieldRaiseException(): void
+    {
+        $this->expectException(\Atk4\Data\Persistence\Sql\ExecuteException::class);
+
+        $master_model = new \Atk4\Data\Model($this->db, [
+            'table' => 'user_join',
+        ]);
+        $master_model->addField('name');
+        $master_model->addField('test_id');
+
+        $joined_model = new \Atk4\Data\Model($this->db, [
+            'table' => 'contact_join',
+            'idField' => 'not_named_id',
+        ]);
+        $joined_model->addField('contact_id');
+        $joined_model->addField('contact_phone');
+
+        $this->createMigrator($master_model)->create();
+        $this->createMigrator($joined_model)->create();
+
+        $master_model->import([
+            [
+                'id' => 1,
+                'name' => 'John',
+                'test_id' => 21,
+            ],
+        ]);
+
+        $joined_model->import([
+            [
+                'not_named_id' => 1,
+                'contact_id' => 21,
+                'contact_phone' => '+123',
+            ],
+        ]);
+
+        $user = new Model($this->db, ['table' => 'user_join']);
+        $user->addField('name');
+        $j = $user->join('contact_join', [
+            'foreignTable'        => 'contact_join',
+            'masterField'         => 'test_id',
+            'foreignField'        => 'contact_id',
+            //'foreignModelIdField' => 'not_named_id',
+        ]);
+        $this->createMigrator()->createForeignKey($j);
+        $j->addField('contact_phone');
+
+        $user = $user->load(1);
+
+        $user->set('name', 'John');
+        $user->set('contact_phone', '+321');
+
+        $user->save();
+    }
+
+    public function testJoinSavingForeignCustomIdField(): void
+    {
+        $master_model = new \Atk4\Data\Model($this->db, [
+            'table' => 'user_join',
+        ]);
+        $master_model->addField('name');
+        $master_model->addField('test_id');
+
+        $joined_model = new \Atk4\Data\Model($this->db, [
+            'table' => 'contact_join',
+            'idField' => 'not_named_id',
+        ]);
+        $joined_model->addField('contact_id');
+        $joined_model->addField('contact_phone');
+
+        $this->createMigrator($master_model)->create();
+        $this->createMigrator($joined_model)->create();
```

Comment:
add foreign key, see other tests how, this will detect correct SQL insert/update/delete order

Reply:
`$joinedModel` created at the start of the test, is only a model and is used to make the import and the final assert.
The `createForeignKey` is added a few lines below on the joined model